src/h5web/visualizations/heatmap/utils.ts

```diff
@@ -17,3 +19,28 @@ export function generateCSSLinearGradient(
 
   return `linear-gradient(to ${direction},${gradientColors})`;
 }
+
+export function getColorScaleDomain(
+  scaleType: ScaleType,
+  values: number[],
+  dataDomain?: Domain,
+  customDomain?: Domain
+): Domain | undefined {
+  const domain = customDomain || dataDomain;
+  if (!domain) {
+    return undefined;
+  }
+
+  // If the scale is not log or if the domain does not cross zero, the domain does not need to be adjusted
+  if (scaleType !== ScaleType.Log || domain[0] * domain[1] > 0) {
+    return domain;
+  }
+
+  const supportedDomain = findDomain(values.filter((x) => x > 0));
```

Comment:
This line filters positive values, but the condition above supports datasets with only negative values as well. The `filter` needs a ternary inside it that chooses which function to apply based on whether `domain[0] * domain[1]` is greater or lower than 0. Perhaps `d3-array` has something to simplify all this logic? I haven't looked.

Reply:
I see your point, but I think this necessitates some thought. Example: if you have a dataset with positive and negative values, you may argue that you want the supported domain to be computed from negative values. Then, how do you choose between the two possibilities ?

From my point of view, I expect the logScale to be used only for positive values, which is why I used this condition (and the same in `LineVis`). That being said, `d3-scale` also accepts a domain of negative values so I allow the log scale to take advantage of that for datasets containing only negatives values, with the condition above.

TL;DR: it raises interesting questions that should be addressed in another MR, ideally conjointly with `LineVis`.

Reply:
When you get to line 39, you're certain that the domain does not span both negative and positive values, it's always either one or the other. If you don't use `x < 0` when the domain is entirely negative, then you'll just see a white heatmap.

Reply:
> based on whether domain[0] * domain[1] is greater or lower than 0

This was wrong, sorry. You can just check the sign of the min.

Reply:
> When you get to line 39, you're certain that the domain does not span both negative and positive values, it's always either one or the other. 
Hm I don't think I get you there. On the contrary, the early return if `domain[0] * domain[1] > 0` (that is min and max have the same sign) ensures that line 39 is executed *if and only if* the domain spans negative and positive values, which is why the domain needs adjustment.

That being said, the discussion is fruitful as thanks to your comments, I discovered an edge case where the domain is `[-X, 0]` which will not trigger the early return and lead indeed to a white heatmap.

Reply:
Oooh right, sorry :sweat: thanks for the clarification - nice catch on the edge case :fishing_pole_and_fish: